cstring.3
.Dd cstring\-VERSION
.Dt CSTRING 3
.Sh NAME
.Nm cstring
.Nd A simple and lightweight string library for C inspired by C++'s
STL string class
.Sh LIBRARY
C String (-lcstring)
.Sh SYNOPSIS
#include <cstring.h>
.Sh DESCRIPTION
.Pp
The
.Nm
library offers a lightweight and fast way to manage
strings with a wide range of useful functions.
.Bl -tag -width Ds
typedef struct _cstring {
        size_t len;             /* string length */
        size_t capacity;        /* string capacity */
        char *str;              /* contents of string */
CSTRING_SORT_ASCENDING 0x01     /* sort in ascending order */
CSTRING_SORT_DESCENDING 0x02    /* sort in descending order */
CSTRING_SORT_CALLBACK 0x04      /* sort using a callback function */
CSTRING_SORT_REST 0x10          /* sort the rest of the array */
} cstring;
.It typedef\ int\ (*cstring_sort_callback)(const void *, const void *);
Used in sort functions.
.El
.Sh USAGE
You must
.Ar always
call the
.Fn cstring_create
and
.Fn cstring_delete
functions whenever you want to make a new instance of
.Ar cstring
and stop using it respectively, in order to not cause any memory
leaks.
.Pp
The recommended way of initializing an empty string is by doing
.Ar cstring foo = cstring_create(CSTRING_INIT_EMPTY)
.Pp
If a function requires a
.Ar char *
you can access the
.Ar .str
field and pass it to the function.
.Sh FUNCTIONS
.Bl -tag -width Ds
.It void Fn cstring_create "const char *s"
Instanciates and initializes a
.Ar cstring
object.
.It void Fn cstring_delete "cstring *cs"
Deallocate string.
.It void Fn cstring_assign "cstring *cs" "const char *s"
Assign a new string to current string.
.It void Fn cstring_insert "cstring *cs" "const char *s" "size_t i"
Insert at a specific index.
.It void Fn cstring_append "cstring *cs" "const char *s"
Append to end of string.
.It void Fn cstring_prepend "cstring *cs" "const char *s"
Prepend to beginning of string.
.It void Fn cstring_erase "cstring *cs" "size_t pos" "size_t len"
Erase a portion of the string.
.It void Fn cstring_erase_matching "cstring *cs" "const char *s"
Erase first match from string.
.It void Fn cstring_erase_all_matching "cstring *cs" "const char *s"
Erase all matches from string.
.It void Fn cstring_trim "cstring *cs" "const char *s"
Trim characters from string.
.It void Fn cstring_push_back "cstring *cs" "char c"
Add a character at the end of the string.
.It void Fn cstring_pop_back "cstring *cs"
Remove the last character in the string.
.It void Fn cstring_replace_char "cstring *cs" "size_t i" "char c"
Replace character at a specific index.
.It void Fn cstring_replace_str "cstring *cs" "const char *s" "size_t pos" "size_t olen"
Replace portion of the string.
.Ar olen
is the length of the old string.
An example use could be:
.br
.Fn cstring_replace_str "&string" "new_word" "cstring_find(&s, old_word)" "old_word_len"
.It cstring Fn cstring_substr "cstring *cs" "size_t pos" "size_t len"
Extract a substring from current string.
.It void Fn cstring_swap "cstring *lhs" "cstring *rhs"
Swap contents of two strings.
.It void Fn cstring_sort "cstring *cs" "size_t len" "int flags" "cstring_sort_callback callback"
Sort an array of cstrings.
If you want to use the builtin comparison pass
.Ar NULL
in the last argument.
In case you want to use your own callback use the
.Ar CSTRING_SORT_CALLBACK
flag and pass your own callback function in the last argument.
You can also combine flags using the bitwise OR operator.
.It void Fn cstring_sort_partial "cstring *cs" "size_t pos" "size_t len" "int flags" "cstring_sort_callback callback"
Like
.Fn cstring_sort
but for specified part of an array.
.It void Fn cstring_sort_chars "cstring *cs" "int flags" "cstring_sort_callback callback"
Sort a cstring's contents.
If you want to use the builtin comparison pass
.Ar NULL
in the last argument.
In case you want to use your own callback use the
.Ar CSTRING_SORT_CALLBACK
flag and pass your own callback function in the last argument.
You can also combine flags using the bitwise OR operator.
.It void Fn cstring_sort_chars_partial "cstring *cs" "size_t pos" "size_t len" "int flags" "cstring_sort_callback callback"
Like
.Fn cstring_sort_chars
but for specified part of string.
.It void Fn cstring_shrink_to_fit "cstring *cs"
Reduce string's capacity to its size.
.It void Fn cstring_clear "cstring *cs"
Erase the whole string.
.It size_t Fn cstring_find "const cstring *cs" "const char *s"
Find first occurence of a pattern in string.
.It size_t Fn cstring_rfind "const cstring *cs" "const char *s"
Find last occurence of a pattern in string.
.It size_t Fn cstring_find_first_of "const cstring *cs" "const char *s"
Find first occurence of specified characters in string.
.It size_t Fn cstring_find_first_not_of "const cstring *cs" "const char *s"
Find the first character that does not match any of the specified characters.
.It size_t Fn cstring_find_last_of "const cstring *cs" "const char *s"
Find last occurence of specified characters in string.
.It size_t Fn cstring_find_last_not_of "const cstring *cs" "const char *s"
Find the last character that does not match any of the specified characters.
.It char Fn cstring_front "const cstring *cs"
Returns the first character of the string.
.It char Fn cstring_back "const cstring *cs"
Returns the last character of the string.
.It int Fn cstring_empty "const cstring *cs"
Check to see if the string is empty.
.It int Fn cstring_starts_with_str "const cstring *cs" "const char *s"
Check to see if string begins with
.Ar s
.It int Fn cstring_ends_with_str "const cstring *cs" "const char *s"
Check to see if string ends with
.Ar s
.It int Fn cstring_starts_with_char "const cstring *cs" "char c"
Check to see if string starts with
.Ar c
.It int Fn cstring_ends_with_char "const cstring *cs" "char c"
Check to see if string ends with
.Ar c
.It void Fn *cstring_data "const cstring *cs"
Get string's content in raw bytes.
.It char Fn *cstring_copy "const char *s"
Make a copy of a given
.Ar const\ char\ *
.It void Fn cstring_resize "cstring *cs" "size_t newcapacity"
Resize the
.Ar str
array inside a given
.Ar cstring
struct.
.It cstring Fn *cstring_getline "FILE *fd" "cstring *cs" "char delim"
Read a line from a
.Ar FILE
stream.
Similar behavior to
.Ar stdio's\ getline
.It int Fn cstring_equal "const cstring *lhs" "const cstring *rhs"
Check if lhs == rhs
.It int Fn cstring_greater "const cstring *lhs" "const cstring *rhs"
Check if lhs > rhs
.It int Fn cstring_greater_or_equal "const cstring *lhs" "const cstring *rhs"
Check if lhs >= rhs
.It int Fn cstring_less "const cstring *lhs" "const cstring *rhs"
Check if lhs < rhs
.It int Fn cstring_less_or_equal "const cstring *lhs" "const cstring *rhs"
Check if lhs <= rhs
.El
.Sh MACROS
.Bl -tag -width Ds
.It Fn CSTRING_OUT_OF_BOUNDS "len" "pos"
Check if
.Ar pos
is out of bounds (pos > len).
.It Fn CSTRING_ARR_LEN "x"
Determine an array's length.
The macro must be called in the same function the array is declared.
.It Fn CSTRING_MALLOC "ptr" "size"
Allocate memory with error cheking.
.El
.Sh CONSTANTS
.Bl -tag -width Ds
.It CSTRING_NPOS
This constant signifies that a pattern hasn't been found inside
the string.
Its value is -1.
.It CSTRING_INIT_EMPTY
Used with
.Fn cstring_create
in case the string is to be initliazed as empty.
.El
.Sh AUTHORS
.An Christos Margiolis Aq Mt christos@margiolis.net


cstring.c
/* See LICENSE file for copyright and license details. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "cstring.h"

#define CSTRING_EXCEEDS_CAPACITY(len, cap) ((len) >= (cap))
#define FIND_OCC(cs, s, f) do {			\
	char *_found;				\
	if ((_found = f(cs->str, (s))) != NULL) \
		return (_found - cs->str);	\
} while (0)

static int cstring_is_one_of(char, const char *);
static void *emalloc(size_t);
static void csfree(cstring *);
static inline int cstring_cmp_greater(const void *, const void *);
static inline int cstring_cmp_less(const void *, const void *);
static inline int cstring_cmp_char_greater(const void *, const void *);
static inline int cstring_cmp_char_less(const void *, const void *);

static int
cstring_is_one_of(char c, const char *s)
{
	for (; *s; s++)
		if (*s == c)
			return (1);
	return (0);
}

static void *
emalloc(size_t nb)
{
	void *p;

	if ((p= malloc(nb)) == NULL) {
		perror("malloc");
		exit(1);
	}
	return (p);
}

static void
csfree(cstring *cs)
{
	if (!cstring_empty(cs))
		free(cs->str);
}

static inline int
cstring_cmp_greater(const void *lhs, const void *rhs)
{
	return (cstring_greater((cstring *)lhs, (cstring *)rhs));
}

static inline int
cstring_cmp_less(const void *lhs, const void *rhs)
{
	return (cstring_less((cstring *)lhs, (cstring *)rhs));
}

static inline int
cstring_cmp_char_greater(const void *lhs, const void *rhs)
{
	return ((*(char *)lhs - *(char *)rhs));
}

static inline int
cstring_cmp_char_less(const void *lhs, const void *rhs)
{
	return (-(*(char *)lhs - *(char *)rhs));
}

cstring
cstring_create(const char *s)
{
	cstring cs;
	cs.len = strlen(s);
	cs.str = cstring_copy(s);
	cstring_resize(&cs, cs.len << 1);
	return (cs);
}

void
cstring_delete(cstring *cs)
{
	csfree(cs);
	cs->str = NULL;
	cs->len = 0;
	cs->capacity = 0;
}

void
cstring_assign(cstring *cs, const char *s)
{
	csfree(cs);
	cs->str = cstring_copy(s);
	cs->len = strlen(s);
	if (CSTRING_EXCEEDS_CAPACITY(cs->len, cs->capacity))
		cstring_resize(cs, cs->len << 1);
}

void
cstring_insert(cstring *cs, const char *s, size_t i)
{
	size_t slen, newlen;
	char *tmp;

	if (!CSTRING_OUT_OF_BOUNDS(cs->len, i) && s != NULL) {
		slen = strlen(s);
		newlen = cs->len + slen;
		tmp = emalloc(newlen + 1);
		memcpy(tmp, cs->str, i);
		memcpy(tmp + i, s, slen);
		memcpy(tmp + i + slen, cs->str + i, newlen - slen - i + 1);
		csfree(cs);
		cs->len = newlen;
		cs->str = tmp;
		cs->str[cs->len] = '\0';
		if (CSTRING_EXCEEDS_CAPACITY(newlen, cs->capacity))
			cstring_resize(cs, newlen << 1);
	}
}

void
cstring_erase(cstring *cs, size_t pos, size_t len)
{
	size_t newlen;
	char *tmp;

	if (!cstring_empty(cs)
	&& (!CSTRING_OUT_OF_BOUNDS(cs->len, pos)
	||  !CSTRING_OUT_OF_BOUNDS(cs->len, len))) {
		newlen = cs->len - len;
		tmp = emalloc(newlen + 1);
		memcpy(tmp, cs->str, pos);
		memcpy(tmp + pos, cs->str + pos + len, cs->len - pos - len);
		csfree(cs); /* Useless check but keep it for consistency */
		cs->len = newlen;
		cs->str = tmp;
		cs->str[cs->len] = '\0';
	}
}

void
cstring_erase_matching(cstring *cs, const char *s)
{
	if (s != NULL)
		cstring_erase(cs, cstring_find(cs, s), strlen(s));
}

void
cstring_erase_all_matching(cstring *cs, const char *s)
{
	size_t len, i;

	if (s != NULL) {
		len = strlen(s);
		i = cstring_find(cs, s);
		for (; i != CSTRING_NPOS; i = cstring_find(cs, s))
			cstring_erase(cs, i, len);
	}
}

void
cstring_trim(cstring *cs, const char *s)
{
	size_t i;
	
	i = cstring_find_first_of(cs, s);
	for (; i != CSTRING_NPOS; i = cstring_find_first_of(cs, s))
		cstring_erase(cs, i, 1);
}

void
cstring_push_back(cstring *cs, char c)
{
	if (CSTRING_EXCEEDS_CAPACITY(cs->len, cs->capacity))
		cstring_resize(cs, cs->len << 1);
	cs->str[cs->len] = c;
	cs->str[++cs->len] = '\0';
}

void
cstring_pop_back(cstring *cs)
{
	if (cs->len > 0)
		cs->str[--cs->len] = '\0';
}

void
cstring_replace_char(cstring *cs, size_t i, char c)
{
	if (!CSTRING_OUT_OF_BOUNDS(cs->len, i))
		cs->str[i] = c;
}

void
cstring_replace_str(cstring *cs, const char *s, size_t pos, size_t olen)
{
	if (!CSTRING_OUT_OF_BOUNDS(cs->len, pos)
	&&  !CSTRING_OUT_OF_BOUNDS(cs->len, olen)) {
		cstring_erase(cs, pos, olen);
		cstring_insert(cs, s, pos);
	}
}

cstring
cstring_substr(const cstring *cs, size_t pos, size_t len)
{
	if (CSTRING_OUT_OF_BOUNDS(cs->len, pos)
	||  CSTRING_OUT_OF_BOUNDS(cs->len, len))
		return (cstring_create(""));
	cstring substr = cstring_create(&cs->str[pos]);
	substr.len = len;
	substr.str[len] = '\0';
	cstring_shrink_to_fit(&substr);
	return (substr);
}

void
cstring_swap(cstring *lhs, cstring *rhs)
{
	cstring tmp = *lhs;
	*lhs = *rhs;
	*rhs = tmp;
}

void
cstring_sort_partial(cstring *cs, size_t pos, size_t len, int flags,
		     cstring_sort_callback cb)
{
	if (flags & CSTRING_SORT_REST ||  CSTRING_OUT_OF_BOUNDS(len, pos + len))
		len -= pos;

	if (flags & CSTRING_SORT_ASCENDING)
		qsort(cs + pos, len, sizeof(cstring), cstring_cmp_greater);
	else if (flags & CSTRING_SORT_DESCENDING)
		qsort(cs + pos, len, sizeof(cstring), cstring_cmp_less);
	else if (flags & CSTRING_SORT_CALLBACK)
		qsort(cs + pos, len, sizeof(cstring), cb);
}

void
cstring_sort_chars_partial(cstring *cs, size_t pos, size_t len, int flags,
			   cstring_sort_callback cb)
{
	if (flags & CSTRING_SORT_REST ||  CSTRING_OUT_OF_BOUNDS(cs->len, pos + len))
		len = cs->len - pos;

	if (flags & CSTRING_SORT_ASCENDING)
		qsort(cs->str + pos, len, sizeof(char), cstring_cmp_char_greater);
	else if (flags & CSTRING_SORT_DESCENDING)
		qsort(cs->str + pos, len, sizeof(char), cstring_cmp_char_less);
	else if (flags & CSTRING_SORT_CALLBACK)
		qsort(cs->str + pos, len, sizeof(char), cb);
}

void
cstring_clear(cstring *cs)
{
	csfree(cs);
	cs->str = emalloc(1);
	cs->str[0] = '\0';
	cs->len = 0;
	cs->capacity = 0;
}

#define CSTRING_CHECK(cs, s)		\
	if (cstring_empty(cs) || !*(s)) \
		return (CSTRING_NPOS)

size_t
cstring_find(const cstring *cs, const char *s)
{
	CSTRING_CHECK(cs, s);
	FIND_OCC(cs, s, strstr);
	return (CSTRING_NPOS);
}

/*SIMPLIFY */
size_t
cstring_rfind(const cstring *cs, const char *s)
{
	size_t idx, slen, i, j;
	int found;

	CSTRING_CHECK(cs, s);
	idx = -1;
	slen = strlen(s);
	for (i = 0; i <= cs->len - slen; i++) {
		found = 1;
		for (j = 0; j < slen; j++) {
			if (cs->str[i + j] != s[j]) {
				found = 0;
				break;
			}
		}
		if (found)
			idx = i;
	}
	return (idx < 0 ? CSTRING_NPOS : idx);
}

size_t
cstring_find_first_of(const cstring *cs, const char *s)
{
	CSTRING_CHECK(cs, s);
	for (; *s; s++) {
		FIND_OCC(cs, *s, strchr);
	}
	return (CSTRING_NPOS);
}

size_t
cstring_find_first_not_of(const cstring *cs, const char *s)
{
	size_t i = 0;

	CSTRING_CHECK(cs, s);
	for (; i < cs->len; i++)
		if (!cstring_is_one_of(cs->str[i], s))
			return (i);
	return (CSTRING_NPOS);
}

size_t
cstring_find_last_of(const cstring *cs, const char *s)
{
	size_t i;

	CSTRING_CHECK(cs, s);
	i = *(s + strlen(s));
	for (; i >= 0; i--) {
		FIND_OCC(cs, s[i], strrchr);
	}
	return (CSTRING_NPOS);
}

size_t
cstring_find_last_not_of(const cstring *cs, const char *s)
{
	size_t i = cs->len;

	CSTRING_CHECK(cs, s);
	for (; i >= 0; i--)
		if (!cstring_is_one_of(cs->str[i], s))
			return (i);
	return (CSTRING_NPOS);
}

#undef CSTR_CHECK

int
cstring_ends_with_str(const cstring *cs, const char *s)
{
	/* TODO: avoid cstring_substr */
	cstring sub;
	size_t slen;
	int found;

	slen = strlen(s);
	sub = cstring_substr(cs, cs->len - slen, slen);
	found = !strcmp(sub.str, s);
	cstring_delete(&sub);
	return (found);
}

char *
cstring_copy(const char *s)
{
	size_t len;
	char *tmp;

	len = strlen(s);
	tmp = emalloc(len + 1);
	memcpy(tmp, s, len + 1);
	tmp[len] = '\0'; /* Add \0 in case `s` didn't have it */
	return (tmp);
}

void
cstring_resize(cstring *cs, size_t newcapacity)
{
	char *tmp;

	tmp = emalloc(newcapacity + 1);
	memcpy(tmp, cs->str, cs->len + 1);
	csfree(cs);
	cs->str = tmp;
	cs->str[cs->len] = '\0';
	cs->capacity = newcapacity;
}

cstring *
cstring_getline(FILE *fd, cstring *cs, char delim)
{
	char c;

	cstring_clear(cs);
	while ((c = fgetc(fd)) != EOF && c != '\n') {
		if (c == delim)
			break;
		else
			cstring_push_back(cs, c);
	}
	return (c == EOF ? NULL : cs);
}
